prow/crier/controller_test.go

```diff
@@ -0,0 +1,183 @@
+	if pj.Status.State == v1.FailureState {
+		*f.passing = true
+	} else {
+		*f.failed = true
+	}
+	return nil, nil
+}
+
+func (f failureReporter) GetName() string {
+	return "failureReporter"
+}
+
+func (f failureReporter) ShouldReport(pj *v1.ProwJob) bool {
+	return pj.Status.State == v1.FailureState
+}
+
+//Fake Informer
+//Sets: The Prow Job Test Cases
+//Asserts: The Controller registers for the queue to be populated
+type fakeInformer struct {
+	isRun        *bool
+	isRegistered *bool
+}
+
+func (f fakeInformer) Get(name string) (*v1.ProwJob, error) {
+	var pj *v1.ProwJob = nil
+	switch name {
+	case "TestJob1":
+		pj = &v1.ProwJob{
+			Spec: v1.ProwJobSpec{
+				Job: "TestJob1",
+			},
+			Status: v1.ProwJobStatus{
+				State: v1.TriggeredState,
+			},
+		}
+	case "TestJob2":
+		pj = &v1.ProwJob{
+			Spec: v1.ProwJobSpec{
+				Job: "TestJob2",
+			},
+			Status: v1.ProwJobStatus{
+				State: v1.FailureState,
+			},
+		}
+	}
+
+	if pj == nil {
+		panic("TODO: return an error")
+	} else {
+		return pj, nil
+	}
+}
+
+func (f fakeInformer) ProwJobs(namespace string) v12.ProwJobNamespaceLister {
+	return f
+}
+
+func (f fakeInformer) Informer() cache.SharedIndexInformer {
+	return f
+}
+
+func (f fakeInformer) Lister() v12.ProwJobLister {
+	return f
+}
+
+func (f fakeInformer) AddEventHandler(handler cache.ResourceEventHandler) {
+	*f.isRegistered = true
+}
+
+func (f fakeInformer) Run(stopCh <-chan struct{}) {
+	*f.isRun = true
+}
+
+func (f fakeInformer) AddEventHandlerWithResyncPeriod(handler cache.ResourceEventHandler, resyncPeriod time.Duration) {
+	*f.isRegistered = true
+}
+
+func (fakeInformer) HasSynced() bool {
+	return true
+}
+
+func (fakeInformer) GetStore() cache.Store                                        { return nil }
+func (fakeInformer) GetController() cache.Controller                              { return nil }
+func (fakeInformer) LastSyncResourceVersion() string                              { return "" }
+func (fakeInformer) AddIndexers(indexers cache.Indexers) error                    { return nil }
+func (fakeInformer) GetIndexer() cache.Indexer                                    { return nil }
+func (fakeInformer) List(selector labels.Selector) (ret []*v1.ProwJob, err error) { return nil, nil }
+
+func TestController_Run(t *testing.T) {
+
+	//Initialization
+	cs := fake.NewSimpleClientset()
+
+	q := kube.RateLimiter("CrierTest")
+	q.Add("TestJob1")
+	q.Add("TestJob2")
+
+	var testResults struct {
+		isRegistered   bool
+		isRun          bool
+		reporterPasses bool
+		reporterFails  bool
+	}
+
+	inf := fakeInformer{}
+	inf.isRegistered = &testResults.isRegistered
+	inf.isRun = &testResults.isRun
+
+	rp := failureReporter{}
+	rp.passing = &testResults.reporterPasses
+	rp.failed = &testResults.reporterFails
+
+	nmwrk := 2
+
+	wg := &sync.WaitGroup{}
+
+	c := NewController(cs, q, inf, rp, nmwrk, wg)
+
+	//Testing
+	stop := make(chan struct{})
+	defer close(stop)
+	go c.Run(stop)
+	time.Sleep(testTimeLimit)
```

Comment:
real sleep in test will be flaky - can we terminate it correctly and check states?

Reply:
What would a less-flaky way of terminating the test after it's processed all of the Prow Jobs, while still failing in a reasonable time if the controller doesn't?

Reply:
I think you can mimic the graceful termination logic from main, feed a stop signal and wait for controller to exit, like https://github.com/kubernetes/test-infra/blob/master/prow/cmd/crier/main.go#L272-L288

Reply:
@krzyzacy The new version of this code keeps the controller running until the queue is emptied, and then checks that everything was handled correctly. A final test is then run that terminates the controller and makes sure it exits gracefully.

I'd appreciate your insight on whether this is a less-flaky solution or not